### PR TITLE
Task10 done

### PR DESCRIPTION
**Pull request by nikmalevich**
Use only the Latin alphabet, since there is no font for the Cyrillic alphabet. Uploading photos is not possible because this is not in the task. For login information, see the Controller class in the _users field.

### CODE REVIEW

**Comment by MaksimViarenich** [comment]
I wanna cry
Снимок экрана 2020-04-29 в 16 18 02

**Comment by MaksimViarenich** [comment]
Can't login. Error in console. Can't do anything
Снимок экрана 2020-04-29 в 16 27 01

**Review by MaksimViarenich — CHANGES_REQUESTED** [review]
Sorry, can't check PR. Can't do anything

**Comment by nikmalevich** [comment]
Can I call you on Skype? Everything starts up and works for me. I don’t know what could be the problem

**Comment by MaksimViarenich** [comment]
> Can I call you on Skype? Everything starts up and works for me. I don’t know what could be the problem

Yes

File: task8/TutterApp/src/main/webapp/index.html [hunk 1]
```diff
@@ -10,65 +10,96 @@
...
-        <script src="resources/js/script.js"></script>
+        <script src="resources/js/model.js"></script>
+        <script src="resources/js/view.js"></script>
+        <script src="resources/js/controller.js"></script>
     </head>
 
     <body>
     <header>
-        <div class="logo">
+        <div id="logo" class="logo">
```
Comment (MaksimViarenich) [inline task8/TutterApp/src/main/webapp/index.html:20]: You don't need add class `logo` to `img` and `p` tags. One class for div is enough.
In CSS you can use selectors `.logo img` and `.logo p `

File: task8/TutterApp/src/main/webapp/index.html [hunk 2]
```diff
@@ -10,65 +10,96 @@
...
             <img src="resources/images/logo.svg" class="logo">
             <p class="logo">Tutter</p>
         </div>
 
-        <div class="log-in-out" id="log-in-out">nikmalevich <button class="log-in-out">Log out</button></div>
+        <div class="log-in-out" id="log-in-out"><span id="username"></span><button class="log-in-out" id="logInOut">Log in</button></div>
```
Comment (MaksimViarenich) [inline task8/TutterApp/src/main/webapp/index.html]: Code style! It's not readable
```
<div class="log-in-out" 
     id="log-in-out">
    <span id="username"></span>
    <button class="log-in-out" 
            id="logInOut">
        Log in
    </button>
</div>
```

File: task8/TutterApp/src/main/webapp/index.html [hunk 3]
```diff
@@ -10,65 +10,96 @@
...
+                <label for="filterName" class="filters">name:</label>
+                <input id="filterName" type="text" class="filters">
+                <label for="filterTag" class="filters">tag:</label>
+                <input id="filterTag" type="text" class="filters">
+                <p class="date">date:</p>
+                <label for="filterDateFrom" class="date-filters">from</label>
+                <input type="date" id="filterDateFrom" name="from" value="2019-01-01" class="date">
+                <label for="filterDateTo" class="date-filters">to</label>
+                <input type="date" id="filterDateTo" name="to" value="2021-01-01" class="date">
+                <button class="add-post" id="addPost">add Post</button>
```
Comment (MaksimViarenich) [inline task8/TutterApp/src/main/webapp/index.html]: the same class for all tags is not needed. Add class `filters` just to `form` tag.
In CSS use selectors: `form.filters` `form.filters p` `form.filters label` `form.filters input`

File: task8/TutterApp/src/main/webapp/resources/js/view.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+    displayPost(post = {}) {
+        let postView = document.importNode(this._postTemplate.content, true);
+
+        this._setPostView(postView, post);
+
+        this._postContainer.insertBefore(postView, this._postContainer.lastElementChild);
+    }
+
+    editPost(id = '', post = {}) {
```
Comment (MaksimViarenich) [inline task8/TutterApp/src/main/webapp/resources/js/view.js]: Don't see usages for this method

File: task8/TutterApp/src/main/webapp/resources/js/model.js [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        if (length !== this._posts.length) {
+            localStorage.removeItem(id);
+
+            return true;
+        }
+
+        return false;
+    }
+
+    addAll(posts = []) {
```
Comment (MaksimViarenich) [inline task8/TutterApp/src/main/webapp/resources/js/model.js]: Don't see usages for this method

File: task8/TutterApp/src/main/webapp/resources/js/model.js [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        return posts.filter(post => !this.add(post));
+    }
+
+    clear() {
```
Comment (MaksimViarenich) [inline task8/TutterApp/src/main/webapp/resources/js/model.js]: Don't see usages for this method

File: task8/TutterApp/src/main/webapp/resources/js/model.js [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        hashTags: val => Array.isArray(val),
+        likes: val => Array.isArray(val)
+    };
+
+    constructor() {
+        let keyPosts = Object.keys(localStorage);
+        this._posts = [];
+
+        for (let i = 0; i < keyPosts.length; i++) {
+            let post = JSON.parse(localStorage.getItem(keyPosts[i]));
```
Comment (MaksimViarenich) [inline task8/TutterApp/src/main/webapp/resources/js/model.js]: Так как в локал сторедже может лежать и другая инфа, а вы обходите всю инфу в локал сторедже, то когда там встречается не JSON, то ваша прилага падает

Из задания: "Заполнение localStorage начальным набором твитов нужно сделать отдельным скриптом и заполнять localStorage при условии что localStorage **на старте пуст**. "

File: task8/TutterApp/src/main/webapp/resources/js/controller.js [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+        let name = formElements[0].value;
+        let hashTags = formElements[1].value;
+        let dateFrom = formElements[2].value;
+        let dateTo = formElements[3].value;
+
+        Controller._curFilter = {};
+
+        if (name !== '') {
+            Controller._curFilter.author = name;
+        }
```
Comment (MaksimViarenich) [inline task8/TutterApp/src/main/webapp/resources/js/controller.js:88]: Minor. Add empty line after `if`

**Review by MaksimViarenich — COMMENTED** [review]
Don't see where you set default number of posts
**Заполнение localStorage начальным набором твитов нужно сделать отдельным скриптом**

**Review by MaksimViarenich — COMMENTED** [review]
When not logged in user will click to any filter and click Enter he will see window for adding new post
Снимок экрана 2020-05-04 в 13 58 37

File: task8/TutterApp/src/main/webapp/resources/js/controller.js [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+        Controller._model.getPage(skip, top, filters).forEach(post => Controller._view.displayPost(post));
+
+        if (Controller._model._curNumFilterPosts <= Controller._curNumVisiblePosts) {
+            Controller._view._seeMoreButton.setAttribute('style', 'display: none');
+        } else {
+            Controller._view._seeMoreButton.setAttribute('style', 'display: block');
+        }
+    }
+
+    static logo(event) {
```
Comment (MaksimViarenich) [inline task8/TutterApp/src/main/webapp/resources/js/controller.js]: Don't see usages for `event` attribute. You can remove it

File: task8/TutterApp/src/main/webapp/resources/js/controller.js [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+        Controller._view._incorrectLogInData.setAttribute('style', 'visibility: hidden');
+        Controller._view._postText.value = '';
+        Controller._view._postTags.value = '';
+        Controller._view._userLogin.value = '';
+        Controller._view._userPassword.value = '';
+
+        Controller.getPage(0, Controller._curNumVisiblePosts, Controller._curFilter);
+    }
+
+    static logInOut(event) {
```
Comment (MaksimViarenich) [inline task8/TutterApp/src/main/webapp/resources/js/controller.js]: Don't see usages for `event` attribute. You can remove it

File: task8/TutterApp/src/main/webapp/resources/js/controller.js [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+    static addPost(event) {
+        Controller._curPost = null;
+
+        Controller._view._mainPage.setAttribute('style', 'display: none');
+        Controller._view._postPage.setAttribute('style', 'display: block');
+
+        event.preventDefault();
+    }
+
+    static donePost(event) {
```
Comment (MaksimViarenich) [inline task8/TutterApp/src/main/webapp/resources/js/controller.js]: Don't see usages for `event` attribute. You can remove it

File: task8/TutterApp/src/main/webapp/resources/js/controller.js [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+                break;
+            case 'far fa-heart like':
+                Controller._model.like(id, Controller._curUser);
+                Controller._view.likePost(id);
+
+                break;
+        }
+    }
+
+    static seeMore(event) {
```
Comment (MaksimViarenich) [inline task8/TutterApp/src/main/webapp/resources/js/controller.js]: Don't see usages for `event` attribute. You can remove it

**Review by MaksimViarenich — CHANGES_REQUESTED** [review]

**Review by MaksimViarenich — APPROVED** [review]

**Review by MaksimViarenich — COMMENTED** [review]
Where is the file for setting a default number of twists?

**Comment by MaksimViarenich** [comment]
> Where is the file for setting a default number of twists?

Заполнение localStorage начальным набором твитов нужно сделать отдельным скриптом

**Pull request merged**